subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformationNodeFactory.java

```diff
@@ -62,14 +62,15 @@ private void collectTransformNodes(ResolvedArtifactSet artifactSet, ImmutableLis
         });
     }
 
-    private TransformationNode getOrCreateInternal(BuildableSingleResolvedArtifactSet singleArtifactSet, List<TransformationStep> transformationChain, ResolvableDependencies resolvableDependencies, ExecutionGraphDependenciesResolver executionGraphDependenciesResolver) {
+    private TransformationNode getOrCreateInternal(BuildableSingleResolvedArtifactSet singleArtifactSet, List<TransformationStep> transformationChain, ResolvableDependencies resolvableDependencies, ExecutionGraphDependenciesResolver executionGraphDependenciesResolver, boolean requiresDependencies) {
         ArtifactTransformKey key = new ArtifactTransformKey(singleArtifactSet.getArtifactId(), transformationChain);
         TransformationNode transformationNode = transformations.get(key);
         if (transformationNode == null) {
             if (transformationChain.size() == 1) {
-                transformationNode = TransformationNode.initial(transformationChain.iterator().next(), singleArtifactSet, resolvableDependencies, executionGraphDependenciesResolver);
+                ArtifactTransformDependenciesProvider dependenciesProvider = DefaultArtifactTransformDependenciesProvider.create(requiresDependencies, singleArtifactSet.getArtifactId(), resolvableDependencies);
+                transformationNode = TransformationNode.initial(transformationChain.get(0), singleArtifactSet, dependenciesProvider, executionGraphDependenciesResolver);
```

Comment:
I don't quite understand how this works. We use a key to identify the transformation. That key does not contain if some transformation in the chain requires dependencies (it cannot, since the initial transformation may not require dependencies, while subsequent will).

So if I have a chain, where only the 3rd transformation requires dependencies, and two tasks, one requiring the output of the 2nd transformation and one requiring the output of the 3rd transformation, then we would create three transformation nodes: for the first, second and third transformation step.

Now, depending on the order in the the task dependencies are discovered, the first two transformation nodes may or may not require dependencies, while the third one always will.

What dependency provider is then passed to the third transformation node? The empty one or the one with the right dependencies?

I think we need a test for this scenario and I also think that the dependency provider needs to be injected per node depending only on the actual transformation. Is this even possible?

Reply:
The alternative is to add `requiresDependencies` to the key, but that also seems wrong - then we would essentially execute the same transformation twice - once with and once without the dependency provider.

Reply:
You are perfectly right, this needs to be fixed in a different way. Will rework that part.